### PR TITLE
Modularity

### PR DESCRIPTION
**Pull request by kkaarreell**
Merging updates from https://github.com/kkaarreell/ci-dnf-stack/tree/modularity
most importantly the modularity code and tests,
updating few other tests to make them functional with latest dnf-3.3 and 3.4
and additionally updates the Dockerfile to use python3.

All tests should be passing when run via 
./dnf-testing --noxfail run
except module-provides-1.feature that fails on dnf-2 due to bug 1623866.

### CODE REVIEW

**Comment by j-mracek** [comment]
module-defaults-1 : dnf works correctly. No idea why why devel profile should be installed

File: Dockerfile [hunk 1]
```diff
@@ -9,12 +9,11 @@ COPY rpms /rpms/
 COPY dnf-docker-test/x509certgen /usr/local/bin
 
 RUN set -x && \
-    echo -e "deltarpm=0\ntsflags=nodocs" >> /etc/dnf/dnf.conf && \
+    echo -e "deltarpm=0" >> /etc/dnf/dnf.conf && \
```
Comment (ignatenkobrain) [inline Dockerfile]: why do we want docs? ;)

File: dnf-docker-test/features/steps/file_steps.py [hunk 2]
```diff
@@ -147,6 +148,54 @@ def step_an_ini_file_filepath_modified_with(ctx, filepath):
...
+            ctx.assertion.assertTrue(conf.has_option(section, key),
+                                     "No such option '%s' in section '%s' in '%s'" % (key, section, filepath))
+            value = settings[key]
+            ini_value = conf.get(section, key)
+            if not extra_value_processing:
+                ctx.assertion.assertEqual(value, ini_value)
+            else:  # an extra processing is enabled
+                if value.startswith('(set)'):
+                    # consider the value to be command or \n separated set of values
+                    value_set = map(str.strip, value[5:].split(","))
```
Comment (ignatenkobrain) [inline dnf-docker-test/features/steps/file_steps.py]: please, don't use `map()`

**Comment by j-mracek** [comment]
module-defaults-install-1 - this is a problem in dnf

**Comment by j-mracek** [comment]
module-disable-2 - Error messages were changed, please can you adjust them to current behavior?

File: dnf-docker-test/features/module-enable-2.feature [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  @setup
+  Scenario: Testing repository Setup
+      Given I run steps from file "modularity-repo-1.setup"
+       When I enable repository "modularityABDE"
+        And I successfully run "dnf makecache"
+        And I successfully run "dnf module enable -y ModuleA:f26"
+        And I successfully run "dnf module enable -y ModuleB:f26"
+
+  Scenario: Enablement of a different stream must be confirmed by the user (no confirmation)
+       When I run "dnf module enable ModuleA:f27"
```
Comment (j-mracek) [inline dnf-docker-test/features/module-enable-2.feature]: Please do not run a command that requires a confirmation waiting for timeout of test step.

**Comment by j-mracek** [comment]
module-enable-2 - must be changed - waiting for timeout is wrong approach

**Comment by j-mracek** [comment]
module-info-1 - dnf work correctly. "dnf module info" should work like "dnf info". Dnf infor works like yum, therefore this is correct behavior.

**Comment by j-mracek** [comment]
module-install-1 - I believe that the problem is in the test.

**Comment by j-mracek** [comment]
module-install-3 - please can you change the test according to new outputs?

**Comment by j-mracek** [comment]
module-list-1 - I belive that dnf work fine here. Please can you also check it or adjust it?

**Comment by j-mracek** [comment]
module-profile-1 - "dnf module profile"  - profile subcommand was removed from dnf. It was only alias for dnf info --profile

**Comment by j-mracek** [comment]
module-provides-1 - may be problem in dnf

**Comment by j-mracek** [comment]
module-remove-1 - I think that dnf works here as expected. Please can you check it?

**Comment by j-mracek** [comment]
module-streams-1 - The streams subcommand was removed from dnf

**Comment by j-mracek** [comment]
module-update-1 - Here dnf changed behavior. It use same strategy like "dnf group upgrade". Onlu opgrades what is installed on system - no installs or remove.

**Comment by j-mracek** [comment]
module-update-2 - Believe that dnf works correctly

**Comment by j-mracek** [comment]
module-update-3 -  Please do not hang command waiting for user action - "dnf module enable ModuleA:f26"

File: dnf-docker-test/features/steps/file_steps.py (hunk 2 above)
Comment (kkaarreell) [reply dnf-docker-test/features/steps/file_steps.py]: what should I use then?

**Comment by kkaarreell** [comment]
I think I will close this pull request as the current approach doesn't make any sense to me. I have ported test thest suite 3 times already to match the currently latest dnf behaviour but since devs are not willing to adopt it right away but are developing new patches in the meantime for me it is impossible to catch up the development of the dnf team that is not reflecting their updates in the test suite itself.

**Comment by kkaarreell** [comment]
I do not understand why did you observe so many failures. What dnf version did you used?
I see less failures.

yours:

module-defaults-1-dnf-3
module-defaults-install-1-dnf-3
module-disable-2-dnf-3
module-enable-2-dnf-3
module-enable-3-dnf-3
module-info-1-dnf-3
module-install-1-dnf-3
module-install-3-dnf-3
module-list-1-dnf-3
module-profile-1-dnf-3
module-provides-1-dnf-3
module-remove-1-dnf-3
module-streams-1-dnf-3
module-update-1-dnf-3
module-update-2-dnf-3
module-update-3-dnf-3
module-update-pkg-1-dnf-3

mine (dnf-3.5.1-1):
module-enable-3-dnf-3
module-provides-1-dnf-3
module-install-3-dnf-3
module-list-1-dnf-3
module-info-1-dnf-3
module-defaults-1-dnf-3
module-enable-2-dnf-3
module-profile-1-dnf-3
module-disable-2-dnf-3
module-streams-1-dnf-3
module-remove-1-dnf-3

File: dnf-docker-test/features/module-enable-2.feature (hunk 3 above)
Comment (kkaarreell) [reply dnf-docker-test/features/module-enable-2.feature]: Why? How would you suggest to test that dnf prompts the user for the confirmation?

File: dnf-docker-test/features/steps/file_steps.py (hunk 2 above)
Comment (ignatenkobrain) [reply dnf-docker-test/features/steps/file_steps.py]: `[v.strip() for v in value[5:].split(",")]` or smth like that

**Comment by j-mracek** [comment]
@no idea, why tre is an difference. Probably problem on my side.

File: Dockerfile (hunk 1 above)
Comment (kkaarreell) [reply Dockerfile]: Now I remember why I did remove it. These tsflags settings is already present in dnf.conf and thus this echo was adding duplicate record which subsequently caused issues in dnf.conf parsing in the ini file step.

**Comment by kkaarreell** [comment]
I did update tests to match the behaviour of  failing tests adjusted to match behavior of dnf-3.5.1-1.git.8534.bf0b69a.fc28.noarch or @xfail referring a bug. 
Again, at this moment only module-provides-1.feature fails on dnf-2 due to bug 1623866.
Please merge this ASAP. :-D

**Comment by j-mracek** [comment]
It looks like that module-provides-1 is failing. Probably little bit randomly, but fails. Please can you solve it?

**Comment by kkaarreell** [comment]
I have added @xfail to that test too, it is failing due to bug 1623866.

**Comment by j-mracek** [comment]
Thanks, all test passed with DNF upstreams

**Pull request merged**